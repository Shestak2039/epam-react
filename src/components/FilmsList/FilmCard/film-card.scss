@import "../../../styles/colors";

.film-wrapper {
  cursor: pointer;
  transition: all 0.5s;

  position: relative;

  &:hover {
    transition: all 0.5s;

    .additional-info-icon {
      position: absolute;
      right: 10px;
      top: 10px;

      width: 30px;
      height: 30px;

      color: $default-text-color;
      background-color: $main-theme;
      border-radius: 100%;

      &:after {
        content: '\2807';

        position: absolute;
        left: 10px;
        top: 1px;

        font-size: 20px;
      }
    }
  }

  .film-info {
    display: flex;
    justify-content: space-between;

    margin-top: 10px;

    color: $default-text-color;

    &-left {
      &-title {
        font-weight: bold;
      }

      &-genres {
        font-size: 14px;
        color: $pink-text-color;

        margin-top: 10px;
      }
    }

    &-right {
      &-year {
        font-size: 15px;

        border: 1px solid $secondary-color;

        padding: 5px;
      }
    }
  }
}
