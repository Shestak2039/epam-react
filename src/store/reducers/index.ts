import { combineReducers } from 'redux';

import filmsReducer from './films-reducer';

export default combineReducers({
    filmsState: filmsReducer
});
