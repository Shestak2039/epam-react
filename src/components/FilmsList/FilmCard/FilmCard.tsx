import React from 'react';

import './film-card.scss';

import { Menu, MenuItem } from '@material-ui/core';
import Modal from '../../common/Modal/Modal';
import FilmForm from '../../FilmForm/FilmForm';
import DeleteFilmModal from '../../DeleteFilmForm/DeleteFilmModal';

import { useModal } from '../../../custom-hooks/modal.hook';
import { FilmCardModel } from '../../../models/film-card.model';

export interface FilmCardProps {
    film: FilmCardModel;
    openPage: (id: string) => void;
}

const FilmCard: React.FunctionComponent<FilmCardProps> = ({film, openPage}) => {
    const [anchorEl, setAnchorEl] = React.useState<Element | ((element: Element) => Element) | null | undefined>(null);

    const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
        setAnchorEl(event.currentTarget);
    };

    const handleClose = () => {
        setAnchorEl(null);
    };

    const [showEditModal, toggleEditModal] = useModal();

    const handleEditButton = () => {
        toggleEditModal();
        handleClose();
    };

    const [showDeleteModal, toggleDeleteModal] = useModal();

    const handleDeleteButton = () => {
        toggleDeleteModal();
        handleClose();
    };

    return (
        <div className="film-wrapper">
            <img src={film.imageUrl} height="450" width="300" onClick={() => openPage(film.id || '')}/>
            <div className="film-info">
                <div className="film-info-left">
                    <div className="film-info-left-title">{film.title}</div>
                    <div className="film-info-left-genres">{film.genres.join(', ')}</div>
                </div>
                <div className="film-info-right">
                    <div className="film-info-right-year">{film.year.split('-')[0]}</div>
                </div>
            </div>
            <div className="additional-info-icon" onClick={handleClick}></div>
            <Menu
                id="simple-menu"
                anchorEl={anchorEl}
                keepMounted
                open={Boolean(anchorEl)}
                onClose={handleClose}
            >
                <MenuItem onClick={handleEditButton}>Edit</MenuItem>
                <MenuItem onClick={handleDeleteButton}>Delete</MenuItem>
            </Menu>
            {showEditModal ? (
                <Modal onClose={toggleEditModal}>
                    <FilmForm closeModal={toggleEditModal} filmInfo={film} isUpdatingForm={true}/>
                </Modal>
            ) : null}
            {showDeleteModal ? (
                <Modal onClose={toggleDeleteModal}>
                    <DeleteFilmModal filmId={film.id || ''} closeModal={toggleDeleteModal}/>
                </Modal>
            ) : null}
        </div>
    )
};

export default FilmCard;
