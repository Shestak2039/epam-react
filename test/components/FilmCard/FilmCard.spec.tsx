import React from 'react';
import { render } from '@testing-library/react';
import FilmCard from '../../../src/components/FilmsList/FilmCard/FilmCard';
import { FilmCardModel } from '../../../src/models/film-card.model';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event/dist';
import { Provider } from 'react-redux';
import store from '../../../src/store/store';

describe('FilmCard', () => {
    const testMovie1: FilmCardModel = {
        id: 'id',
        title: 'title',
        genres: [],
        year: 'year',
        imageUrl: 'imageUrl',
        rating: 0,
        durationInMinutes: 0,
        overview: 'overview',
    };
    
    test('renders FilmCard snapshot', () => {
        const { asFragment, getByTestId } = render(<Provider store={store}><BrowserRouter><FilmCard film={testMovie1}/></BrowserRouter></Provider>);

        userEvent.click(getByTestId('additional-icon'));
        userEvent.click(getByTestId('edit-button'));
        userEvent.click(getByTestId('delete-button'));
        
        expect(asFragment()).toMatchSnapshot();
    });
});
