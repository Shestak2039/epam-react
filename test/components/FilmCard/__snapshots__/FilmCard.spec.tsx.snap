// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmCard renders FilmCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="film-wrapper"
  >
    <a
      href="/film/id"
    >
      <img
        height="450"
        src="imageUrl"
        width="300"
      />
    </a>
    <div
      class="film-info"
    >
      <div
        class="film-info-left"
      >
        <div
          class="film-info-left-title"
        >
          title
        </div>
        <div
          class="film-info-left-genres"
        />
      </div>
      <div
        class="film-info-right"
      >
        <div
          class="film-info-right-year"
        >
          year
        </div>
      </div>
    </div>
    <div
      class="additional-info-icon"
      data-testid="additional-icon"
    />
  </div>
</DocumentFragment>
`;
